buster/documents/sqlite/documents.py

```diff
@@ -0,0 +1,157 @@
+import itertools
+import sqlite3
+import warnings
+import zlib
+from pathlib import Path
+from typing import Iterable, NamedTuple
+
+import numpy as np
+import pandas as pd
+
+import buster.documents.sqlite.schema as schema
+from buster.documents.base import DocumentsManager
+
+
+class Section(NamedTuple):
+    title: str
+    url: str
+    content: str
+    parent: int | None = None
+    type: str = "section"
+
+
+class Chunk(NamedTuple):
+    content: str
+    n_tokens: int
+    emb: np.ndarray
+
+
+class DocumentsDB(DocumentsManager):
+    """Simple SQLite database for storing documents and questions/answers.
+
+    The database is just a file on disk. It can store documents from different sources, and it can store multiple versions of the same document (e.g. if the document is updated).
+    Questions/answers refer to the version of the document that was used at the time.
+
+    Example:
+        >>> db = DocumentsDB("/path/to/the/db.db")
+        >>> db.write_documents("source", df)  # df is a DataFrame containing the documents from a given source, obtained e.g. by using buster.docparser.generate_embeddings
+        >>> df = db.get_documents("source")
+    """
+
+    def __init__(self, db_path: sqlite3.Connection | str):
+        if isinstance(db_path, (str, Path)):
+            self.db_path = db_path
+            self.conn = sqlite3.connect(db_path, detect_types=sqlite3.PARSE_DECLTYPES)
+        else:
+            self.db_path = None
+            self.conn = db_path
+        schema.initialize_db(self.conn)
+        schema.setup_db(self.conn)
+
+    def __del__(self):
+        if self.db_path is not None:
+            self.conn.close()
+
+    def get_current_version(self, source: str) -> tuple[int, int]:
+        """Get the current version of a source."""
+        cur = self.conn.execute("SELECT source, version FROM latest_version WHERE name = ?", (source,))
+        row = cur.fetchone()
+        if row is None:
+            raise KeyError(f'"{source}" is not a known source')
+        sid, vid = row
+        return sid, vid
+
+    def get_source(self, source: str) -> int:
+        """Get the id of a source."""
+        cur = self.conn.execute("SELECT id FROM sources WHERE name = ?", (source,))
+        row = cur.fetchone()
+        if row is not None:
+            (sid,) = row
+        else:
+            cur = self.conn.execute("INSERT INTO sources (name) VALUES (?)", (source,))
```

Comment:
Good point. I say we let it like that for now, because I'm changing the way we insert sections as part of the parsing, but the retriever should not create a new version, ever. My assumption is that we will retrieve using the view, but this method is dangerous as-is.